Draft mks ldp (#5744)
ovh/docs: pages/public_cloud/containers_orchestration/managed_kubernetes/forwarding-audit-logs-to-logs-data-platform/guide.en-gb.md

---
title: Managed Kubernetes Service Audit Logs Forwarding
excerpt: 'Find out how to forward Audit logs from an OVH Managed Kubernetes Service cluster to Logs Data Platform'
updated: 2024-30-01
---

 
## Objective
In this guide, you will learn how to enable the forwarding of audit logs from your OVHcloud Managed Kubernetes Service (MKS) cluster to Logs Data Platform (LDP), a platform that helps you store, archive, query and visualize your logs.
If you want to discover Logs Data Platform before reading this guide, please refer to [Log Data Platform introduction guide](https://help.ovhcloud.com/csm/en-gb-logs-data-platform-introduction?id=kb_article_view&sysparm_article=KB0058312).

## Glossary
- **Logs Data Platform:** a fully managed and secured log management platform by OVHcloud. Find more information on the [Logs Data Platform service page](https://www.ovhcloud.com/en-ie/logs-data-platform/).

> - **Data Stream:** a logical partition of logs that you create in an LDP account that you will use when ingesting, visualizing or querying your logs. Multiple sources can be stored in the same data stream, and it is the unit for defining a logs pipeline (retention policy, archiving, live streaming...), access rights and alert policies.

Reviewer: ```suggestion
- **Data Stream:** a logical partition of logs that you create in a LDP account that you will use when ingesting, visualizing or querying your logs. Multiple sources can be stored in the same data stream, and it is the unit for defining a logs pipeline (retention policy, archiving, live streaming...), access rights and alert policies.
```

> - **Logs forwarding:** feature integrated in an OVHcloud product to ingest logs from its services to a *Data Stream* of an LDP account in the same OVHcloud account. The feature has to be enabled by the customer and per service.

Reviewer: ```suggestion
- **Logs forwarding:** feature integrated in an OVHcloud product to ingest logs from its services to a *Data Stream* of a LDP account in the same OVHcloud account. The feature has to be enabled by the customer and per service.
```

- **Logs forwarding Subscription:** when enabling the logs forwarding for a given OVHcloud service to a given LDP *Data Stream*, a *Subscription* is created and attached to the *Data Stream* for further management by the customer.

> - **Request Stage and Audit Level:**  audit records begin their lifecycle inside the kube-apiserver component. Each request on each stage of its execution generates an audit event, which is then pre-processed according to the policy defined by the OVHcloud Managed Kubernetes Service. This policy defines which audit events as well as which audit level (meaning which events data) are forwarded through the Audit logs of your Kubernetes cluster. For more details, refer to the Kubernetes Auditing documentation.  

Reviewer 2: In "For more details, refer to the Kubernetes Auditing documentatio", maybe add a link to https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/
Edit: link is already given later. Maybe give it here first, as you mention it here first ?

Author: ```suggestion
- **Request Stage and Audit Level:**  audit records begin their lifecycle inside the kube-apiserver component. Each request on each stage of its execution generates an audit event, which is then pre-processed according to the policy defined by the OVHcloud Managed Kubernetes Service. This policy defines which audit events as well as which audit level (meaning which events data) are forwarded through the Audit logs of your Kubernetes cluster. For more details, refer to the [Kubernetes Auditing documentation](https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/).  
```

## Requirements
To follow this guide, you will need:
-	a Logs Data Platform (LDP) account with at least one active *Stream* configured. This guide will walk you through all the necessary steps.
-	an up-and-running Managed Kubernetes Service (MKS) cluster. This guide will walk you through the steps as well.

> Both LDP account and MKS cluster must belong to the same OVHcloud accounts

Reviewer 2: Not sure for this one, but maybe `account` instead of `accounts`

Author: ```suggestion
Both LDP account and MKS cluster must belong to the same OVHcloud account.
```

> Both LDP account and MKS cluster must belong to the same OVHcloud accounts

Reviewer: ```suggestion
Both LDP account and MKS cluster must belong to the same OVHcloud accounts.
```

Author: fixed

> Managed Kubernetes Audit Logs does NOT include logs of your applications running on your Kubernetes pods. If you want to retrieve your data plan logs in one of your Logs Data Platform’s data stream follow the guide: [Pushing logs from a Kubernetes cluster to Logs Data Platform using Fluent Bit](https://help.ovhcloud.com/csm/en-gb-logs-data-platform-kubernetes-fluent-bit?id=kb_article_view&sysparm_article=KB0050055)

The OVHcloud Managed Kubernetes Service has defined audit policy, enabling you to retrieve logs pertaining to:
-	requests to authorization resources to help troubleshoot authentication issues
-	configmap and secret changes in all namespaces at the Metadata audit level
-	changes to resources at RequestResponse level (maximum verbosity) to create, patch, update and delete verbs
-	changes to resources at Request level for other verbs
-	all other requests at the Metadata level

Note:
-	No logs for requests by the "system:kube-proxy" on endpoints or services
-	Low level of verbosity (level Metadata) for endpoints containing sensitive data like tokenreview
-	No logs for any request in the RequestReceived stage
-	No logs for health checks and requests for apiserver metrics

> 

Reviewer: ```suggestion
```

Above, you will find an example of an audit log generated by a Kubernetes cluster. *Note the example is not exhaustive.*

```json
{
   "kind":"Event",
   "apiVersion":"audit.k8s.io/v1",
   "level":"Metadata",
   "auditID":"XXXXXXX-XXXX-XXXX-XXXX-XXXXXXXX",
   "stage":"ResponseStarted",
   "requestURI":"/api/v1/namespaces/searching-worker/configmaps?......",
   "verb":"watch",
   "user":{
      "username":"system:node:search-workers-node-81097c",
      "groups":[
         "system:nodes",
         "system:authenticated"
      ]
   },
   "sourceIPs":[
      "10.10.0.43"
   ],
   "userAgent":"kubelet/v1.26.4 (linux/amd64) kubernetes/f89670c",
   "objectRef":{
      "resource":"configmaps",
      "namespace":"searching-worker",
      "name":"kube-root-ca.crt",
      "apiVersion":"v1"
   },
   "responseStatus":{
      "metadata":{

      },
      "code":200
   },
   "requestReceivedTimestamp":"2024-01-01T11:59:06.428015Z",
   "stageTimestamp":"2024-01-01T11:59:06.428962Z",
   "annotations":{
      "authorization.k8s.io/decision":"allow",
      "authorization.k8s.io/reason":""
   }
}
```

## How to enable Audit Log Forwarding
### Using the Manager

#### Create and see existing streams

>  You can find your streamId, in Logs Data Platform control panel of your account in Data stream” pan when editing a stream:

Reviewer 2: Data stream” → ”Data stream”

Also a small detail: elsewhere in OVH doc we often talk about "[tab](https://github.com/search?q=repo%3Aovh%2Fdocs+%22+tab+%22&type=code)" and not about "[pan](https://github.com/search?q=repo%3Aovh%2Fdocs+%22+pan+%22&type=code)"

Author: ```suggestion
 You can find your streamId, in Logs Data Platform control panel of your account in **Data Stream**” tab when editing a stream:
```

>  -	Got to the “Data Stream” page of your Logs Data Platform account and “Edit” the target Data Stream

Reviewer: ```suggestion
 -	Got to the “Data Stream” page of your Logs Data Platform account and “Edit” the target Data Stream.
```

 https://www.ovh.com/manager/#/dedicated/dbaas/logs/{ldp-account-id}/streams/home

 -	Copy the streamId page of your Logs Data Platform account.

 https://www.ovh.com/manager/#/dedicated/dbaas/logs/{ldp-account-id}/streams/{streamid}/edit

> ### Using APIs

Reviewer 3: Is it possible through Terraform?

Author: not yet, I will update the doc when available

You will have to define the targeted *Stream* of one of your LDP account on which you want your logs to be forwarded to. The enablement of the forwarding will create a subscription for this stream id.
Note that the forwarding activation is free of charge, but you will be charged for the usage of Logs Data Platform service as per standard price plan. For LDP pricing refer to [Logs Data Platform product page](https://www.ovhcloud.com/en-ie/logs-data-platform/).

> You can retrieve the API specification in [OVH API Portal](https://eu.api.ovh.com/console-preview/?section=%2Fdbaas%2Flogs&branch=v1#post-/dbaas/logs/-serviceName-/output/graylog/stream)

Reviewer: ```suggestion
You can retrieve the API specifications in [OVH API Portal](https://eu.api.ovh.com/console-preview/?section=%2Fdbaas%2Flogs&branch=v1#post-/dbaas/logs/-serviceName-/output/graylog/stream)
```

> ###### 1/ Retrieve your target Stream (and ID):

Reviewer: ```suggestion
###### 1/ Retrieve your target Stream (and ID)
```

> List data stream of you Logs Data Platform account ([API reference](https://eu.api.ovh.com/console-preview/?section=%2Fdbaas%2Flogs&branch=v1#get-/dbaas/logs/-serviceName-/output/graylog/stream)):

Reviewer: ```suggestion
List data streams of you Logs Data Platform account ([API reference](https://eu.api.ovh.com/console-preview/?section=%2Fdbaas%2Flogs&branch=v1#get-/dbaas/logs/-serviceName-/output/graylog/stream)):
```

```shell
GET /dbaas/logs/{serviceName}/output/graylog/stream
```
Get the details of a data stream ([API reference](https://eu.api.ovh.com/console-preview/?section=%2Fdbaas%2Flogs&branch=v1#get-/dbaas/logs/-serviceName-/output/graylog/stream/-streamId-)):
```shell

> GET   /dbaas/logs/{serviceName}/output/graylog/stream/{streamId}

Reviewer: ```suggestion
GET /dbaas/logs/{serviceName}/output/graylog/stream/{streamId}
```

```   
###### 2/ Create your subscription

As per below example, the POST request has a payload containing a streamId, which is the target data stream of your LDP account where you aim your Kubernetes cluster Audit logs will be forwarded. You also need to specifiy the 'kind' of log you want to forward, note that currenlty the only supported value for Managed Kubernetes Service is 'audit' (you can find available kinds using the [dedicated API](https://eu.api.ovh.com/console-preview/?section=%2Fcloud&branch=v1#get-/cloud/project/-serviceName-/capabilities/kube/log/kind) call) :
```shell
POST /cloud/project/{serviceName}/kube/{kubeId}/log/subscription
{

>   "kind": "string",   //Currently the only supported value is 'audit'

Reviewer: ```suggestion
  "kind": "string", // Currently the only supported value is 'audit'.
```

>   "streamId": "18d602ec-af40-4000-8e59-41ecc8c23f80" //the streamID of the targeted Stream

Reviewer: ```suggestion
  "streamId": "18d602ec-af40-4000-8e59-41ecc8c23f80" // The streamID of the targeted Stream.
```

}
```

> You will get in response an operationId

Reviewer: ```suggestion
You will get in response an operationId:
```

```shell
{

> "operationId": "18d3aecd-bae0-4000-8558-4b60e496bb80",

Reviewer 2: You can indent these 2 lines

Also can you use another UUID for this `operationId` ? 
If I try to read it very fast (with my eyes almost closed), it seems that this `operationId` "`18d3...b80`" looks very similar to the `streamId` from the payload above ("`18d6..f80"`"). At first I thought it was the same, and I was confused.
Let's pick `f550aa1c-89ab-4b1a-81ae-4fba4959966f` for instance :)

Author: fixed

> "operationId": "18d3aecd-bae0-4000-8558-4b60e496bb80",

Author: ```suggestion
  "operationId": "f550aa1c-89ab-4b1a-81ae-4fba4959966f",
```

> "serviceName": "string"

Author: ```suggestion
  "serviceName": "string"
```

}
```

> So you can use it to retrieve the subscriptionid further management purposes using Logs Data Platform read operation endpoint:

Reviewer 2: subscriptionid → `subscriptionId`

Here you describe how to use the `operationId` to get you `subscriptionId`, but you directly give the MKS endpoint. That's confusing

Instead you can describe how to follow this operation through [LDP operation route](https://eu.api.ovh.com/console-preview/?section=%2Fdbaas%2Flogs&branch=v1#get-/dbaas/logs/-serviceName-/operation/-operationId-), which will give you the `subscriptionId` once succeed

And then continue wit this `GET /cloud/project/{serviceName}/kube/{kubeId}/log/subscription/{subscriptionId}` description

```shell
GET /cloud/project/{serviceName}/kube/{kubeId}/log/subscription/{subscriptionId}
{
  "createdAt": "2024-01-31T15:45:25.286Z",
  "kind": "string",
  "resource": {
    "name": "string",
    "type": "string"
  },
  "serviceName": "string",
  "streamId": "string",
  "subscriptionId": "18d60324-b260-4000-83db-b484f4db6e80",
  "updatedAt": "2024-01-31T15:45:25.286Z"
}
```

## How to use your Kubernetes Audit logs?
Now that your Kubernetes instance Audit logs are ingested and stored in your Logs Data Platform data stream, you can query your logs and build dashboards to have a graphical representation of your logs using web-based UI of Graylog.

> -	On the OVHcloud Mananger retrieve the admin user and its password in your Logs Data Platform account home page.

Reviewer 2: `LDP username` instead of `admin user`. You can also specify that it starts with `logs-xxxx` (people are often confused between LDP username & their OVH nic)

You can link this doc: [Quick start for Logs Data Platform](https://help.ovhcloud.com/csm/en-ie-logs-data-platform-quick-start?id=kb_article_view&sysparm_article=KB0055819)

> -	On the OVHcloud Mananger retrieve the admin user and its password in your Logs Data Platform account home page.

Reviewer: ```suggestion
-	Through the OVHcloud Manager, retrieve the admin user and its password in your Logs Data Platform account home page.
```

Author: ```suggestion
-	Through the OVHcloud Manager, retrieve the LDP username (ex: logs-xxxx) and its password in your Logs Data Platform account home page. You can refer to [Quick start for Logs Data Platform](https://help.ovhcloud.com/csm/en-ie-logs-data-platform-quick-start?id=kb_article_view&sysparm_article=KB0055819) documentation.
```

> -	Open the Graylog web-ui. You can retrieve le link in your account home page or using your Access point depending of your account region (for example: Gravelines regions https://gra1.logs.ovh.com/ )

Reviewer: ```suggestion
-	Open the Graylog web-ui. You can retrieve the link in your account home page or using your Access point depending of your account region (for example: Gravelines region https://gra1.logs.ovh.com/ )
```

> -	Login Graylog using your Logs Data Platform service name and Password

Reviewer 2: Login Graylog → Login to Graylog
service name → LDP username

Author: ```suggestion
-	Log into Graylog using your Logs Data Platform Username and Password
```

> -	Search through your logs across the data stream of your Logs Data Platform account. You can refer to Graylog writing search queries documentation for details on search syntax.

Reviewer 2: For "You can refer to Graylog writing search queries documentation for details on search syntax", you can link the official documentation: https://go2docs.graylog.org/4-x/making_sense_of_your_log_data/writing_search_queries.html

As we are "stuck" in Graylog v4.x, this link won't change

Author: ```suggestion
-	Search through your logs across the data stream of your Logs Data Platform account. You can refer to [Graylog writing search queries documentation](https://go2docs.graylog.org/4-x/making_sense_of_your_log_data/writing_search_queries.html) for details on search syntax.
```

For more details about how to use your logs with Logs Data Platform, including:
-	how to setup alerts,
-	view the logs in real time through a WebSocket,
-	build visualization with OpenSearch Dashboards,
-	integrate with OpenSearch API
-	to connect with Grafana

> Refer to following documentation: [Logs Data Platform - Visualizing, querying and exploiting your logs](https://help.ovhcloud.com/csm/fr-documentation-observability-logs-data-platform-visualizing-querying-exploiting?id=kb_browse_cat&kb_id=3d4a8129a884a950f07829d7d5c75243&kb_category=00fe0f8829ffe1141e11171b7d895c8c).

Reviewer: ```suggestion
Refer to the following documentation: [Logs Data Platform - Visualizing, querying and exploiting your logs](https://help.ovhcloud.com/csm/fr-documentation-observability-logs-data-platform-visualizing-querying-exploiting?id=kb_browse_cat&kb_id=3d4a8129a884a950f07829d7d5c75243&kb_category=00fe0f8829ffe1141e11171b7d895c8c).
```

   "user":{
      "username":"system:node:search-workers-node-81097c",
      "groups":[
         "system:nodes",
         "system:authenticated"
      ]
   },
   "sourceIPs":[
      "10.10.0.43"
   ],
   "userAgent":"kubelet/v1.26.4 (linux/amd64) kubernetes/f89670c",
   "objectRef":{
      "resource":"configmaps",
      "namespace":"searching-worker",
      "name":"kube-root-ca.crt",
      "apiVersion":"v1"
   },
   "responseStatus":{
      "metadata":{

      },
      "code":200
   },
   "requestReceivedTimestamp":"2024-01-01T11:59:06.428015Z",
   "stageTimestamp":"2024-01-01T11:59:06.428962Z",
   "annotations":{
      "authorization.k8s.io/decision":"allow",
      "authorization.k8s.io/reason":""
   }
}
```

## Instructions

How to enable Audit Log Forwarding

### Enabling Audit Log Forwarding using the OVHcloud Control Panel

#### Create and see existing streams

You can find your streamId, in the Logs Data Platform section of the [OVHcloud control panel](https://www.ovh.com/auth/?action=gotomanager&from=https://www.ovh.co.uk/&ovhSubsidiary=GB).

- Go to the `Data Stream`{.action} tab of your Logs Data Platform account and `Edit`{.action} the target Data Stream.

> https://www.ovh.com/manager/#/dedicated/dbaas/logs/{ldp-account-id}/streams/home

Reviewer 4: If this URL is meant to be clickable, it should be formatted as such
OR I suggest replacing those with screenshots from the OVHcloud Control Panel.

Author: Remplaced by sceenshots (when available) or instructions. We will add more screenshots when EN translation will be available on the OVH Control Panel

- Copy the `streamId` page of your Logs Data Platform account.

> https://www.ovh.com/manager/#/dedicated/dbaas/logs/{ldp-account-id}/streams/{streamid}/edit

Reviewer 4: Same comment as above (line 111)

### Enabling Audit Log Forwarding using APIs

You will have to define the targeted *Stream* of one of your LDP account on which you want your logs to be forwarded to. The enablement of the forwarding will create a subscription for this stream id.
Note that the forwarding activation is free of charge, but you will be charged for the usage of the Logs Data Platform service as per standard price plan. For LDP pricing, refer to the [Logs Data Platform product page](https://www.ovhcloud.com/en-gb/logs-data-platform/).

You can retrieve the API specifications in the [OVH API Portal](https://eu.api.ovh.com/console-preview/?section=%2Fdbaas%2Flogs&branch=v1#post-/dbaas/logs/-serviceName-/output/graylog/stream)

#### Step 1 - Retrieve your target Stream (and ID)

List data streams of you Logs Data Platform account:

[API reference](https://eu.api.ovh.com/console-preview/?section=%2Fdbaas%2Flogs&branch=v1#get-/dbaas/logs/-serviceName-/output/graylog/stream)

```shell

> GET /dbaas/logs/{serviceName}/output/graylog/stream

Reviewer 4: Why not using our dedicated API calls markdown formats for all the API calls?

Author: fixed
